### DISCUSSION TITLE
Why can't I disable these two analyzer rules?

### DISCUSSION THREAD

**Discussion by bichuga** [dotnet/roslyn#59364] [Q&A] [answered]
I have a project in my enlistment where Roslyn analyzers CA5390 and CA5403 both add 23 minutes of build time so I'm trying to disable them. I created a `.editorconfig` file in the project directory to disable them, but when I build, the analyzers are still running!  I created the exact same `.editorconfig` in a different project and it turns off the rules as expected.

I can tell the rules are running by adding `/reportanalyzer` to my command line and both show up at the top.

I've tried creating a preprocessing log `/pp:pp.log` and diff'd them. Besides the csproj itself, the pp.log is identical. 

This is the content of my `.editorconfig`:
```
# disable CA5390 and CA5403 due to ~1400 seconds of added build time
# 1405.528s Microsoft.NetCore.Analyzers.Security.DoNotHardCodeEncryptionKey (CA5390)
# 1377.102s Microsoft.NetCore.Analyzers.Security.DoNotHardCodeCertificate (CA5403)
[*.{cs,vb}]
dotnet_diagnostic.CA5403.severity=none
dotnet_diagnostic.CA5390.severity=none
```

If I load the project in visual studio and view properties on the project, it says these two analyzers are Disabled. I'm out of ideas. How can I find why these rules are running?

I'm running version 3.3.2 of each of these analyzer packages. These are .net framework projects (not .netsdk style). There are no .editorconfig files higher up in the tree for either project.

**Comment by sharwell** [comment, marked as answer]
Try disabling the rules via **.globalconfig** instead of via **.editorconfig**.

https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/configuration-files#global-analyzerconfig

**Reply by bichuga** [reply]
thanks!

**Reply by Youssef1313** [reply]
@sharwell Is it known why `.editorconfig` may not work here? Probably some interference if a whole category is enabled or something? (setting severities for CAxxxx almost always surprises me and usually hard to explain due to the magic behind `AnalysisMode`, etc)

**Reply by sharwell** [reply]
**.editorconfig** disables a rule only in files where it's applied. It does this through two steps:

1. It disables analysis if possible
2. If something is reported, the warning severity is lowered to none and the diagnostic suppressed

These two rules are special because they require analysis of all files (including generated files) in order to correctly report diagnostics anywhere in the project:
https://github.com/dotnet/roslyn-analyzers/blob/65680a5750b0778ab869b02a70cc5e675410201e/src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Security/SourceTriggeredTaintedDataAnalyzerBase.cs#L49

Due to this configuration, optimization (1) is impossible to apply. So the rule still runs, it just doesn't report anything.

**.globalconfig** applies a broader rule that applies to all sources in a project, including ones that **.editorconfig** cannot (e.g. files outside the main directory structure).

I may be wrong on one or more details here, but the actual behavior is "something like this".

**Reply by Youssef1313** [reply]
@sharwell Almost all generated files will end in `.cs` (specifically `.g.cs`, `.generated.cs`, etc), so why `[*.{cs,vb}]` didn't cover them?

I never imagined that analyzing and reporting in generated code can affect configuring rules. This is really good to know!

**Reply by sharwell** [reply]
@Youssef1313 I'm not sure source generated files are considered to be in the path of the **.editorconfig**. It's also possible to have files linked in from other locations, including but not limited to NuGet cache directories.

**Reply by Youssef1313** [reply]
Ah that makes sense. Thanks @sharwell!
